### PR TITLE
Add Logging Library SDK Prototype Specification

### PR DESCRIPTION
**Pull request by tigrannajaryan**
This is a draft proposal for OpenTelemetry Logging Library SDK specification.
The purpose of his proposal is to lay out the initial understanding of
what Logging Library SDKs will look like.

We would like to create prototypes based on this proposal in a few languages.
The learnings from prototypes will then allow us to refine this proposal and
eventually make the proposed approach part of the OpenTelemetry specification.

The approach proposed in this OTEP is not intended to be merged into the
OpenTelemetry specification after the OTEP itself is approved. We want to have
successful prototypes first. This OTEP is first and foremost a specification and
guidelines on how to create the prototypes.

The specification defines how the OpenTelemetry Logging Library SDK exposes its
functionality to authors of extensions to language-specific 3rd party logging
libraries and to end users that want to produce logs in
[OpenTelemetry manner](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/overview.md).

The specification defines SDK elements that to some extent mirror the
OpenTelemetry
[Trace SDK](https://github.com/open-telemetry/opentelemetry-specification/tree/main/specification/trace).
This ensures uniformity and consistency of the OpenTelemetry specification and
of the implementations across traces and logs. For additional clarity the
definitions in this document refer to the Trace analogs where appropriate.

The descriptions of interfaces and methods in this document are intentionally
very brief. Detailed and precise descriptions will be borrowed from Trace
specification when this OTEP is submitted as a Log specification PR. I kept the
descriptions as short as possible in this document to make reviewing it easy and
fast.

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
@pmm-sumo for some reason I cannot add you as a reviewer, please review.

**Comment by tigrannajaryan** [comment]
Note: I am out tomorrow, won't be able to respond to comments until the day after.

File: text/logs/0000-logging-library-sdk.md
Comment (pmm-sumo) [inline text/logs/0000-logging-library-sdk.md:206]: I wonder if some users would want to have OTLP/Console exporter as well (which could e.g. print JSON-formatted records)

**Review by pmm-sumo — APPROVED** [review]

File: text/logs/0000-logging-library-sdk.md
Comment (veera83372) [reply text/logs/0000-logging-library-sdk.md:206]: I too think we should provide OTLP/Console but, it would be OTLP/File(stdout) may we could add that wrapper too.

Comment (tigrannajaryan) [reply text/logs/0000-logging-library-sdk.md:206]: Yes, the was the intent. OTLP/File can output to stdout as an option. We may also provide a wrapper to make it trivial.

File: text/logs/0000-logging-library-sdk.md
Comment (tigrannajaryan) [reply text/logs/0000-logging-library-sdk.md:86]: Yes, I agree. If we don't manage to have the answer as a result of PR discussion let's make sure the prototyping helps us to get the answer.

**Review by bogdandrutu — APPROVED** [review]
For a prototype phase these specifications are good

File: text/logs/0000-logging-library-sdk.md
Comment (bogdandrutu) [inline text/logs/0000-logging-library-sdk.md:4]: **Status**

Initial draft specification to create prototypes of OpenTelemetry Logging
Library SDKs.

**Prototypes**

* x
* y
* z

Comment (tigrannajaryan) [reply text/logs/0000-logging-library-sdk.md]: I added "Status" to be explicit.

Will add a list of prototypes later when we have them.

**Comment by tigrannajaryan** [comment]
@open-telemetry/python-approvers I think there was some recent log-related work done in Python library. Please review this proposal.

**Comment by tigrannajaryan** [comment]
@open-telemetry/cpp-approvers I believe we have a Log prototype in C++ library. Please review this proposal and see if it fits what you wanted to do in C++.

File: text/logs/0150-logging-library-sdk.md
Comment (perk-sumo) [inline text/logs/0150-logging-library-sdk.md:122]: Something is off with this sentence.
Should the `How to Create Log4J Style Appender` be a header for this section or sth?

Comment (tigrannajaryan) [reply text/logs/0150-logging-library-sdk.md]: Yes, incorrect formatting. Fixed now, thanks.

**Review by zenmoto — APPROVED** [review]

**Review by djaglowski — APPROVED** [review]

File: text/logs/0150-logging-library-sdk.md
Comment (jkwatson) [inline text/logs/0150-logging-library-sdk.md:16]:
```suggestion
The learnings from the prototypes will then allow us to refine this proposal and
```

File: text/logs/0150-logging-library-sdk.md
Comment (jkwatson) [inline text/logs/0150-logging-library-sdk.md:26]:
```suggestion
libraries and to end users that want to produce logs in the
```

File: text/logs/0150-logging-library-sdk.md
Comment (jkwatson) [inline text/logs/0150-logging-library-sdk.md:78]: Note: Spring Cloud Sleuth currently enables appending Sleuth Baggage to log records via the slf4j MDC, so there is definitely prior art for baggage being appended to the logging MDC.

See for documentation of this feature: https://docs.spring.io/spring-cloud-sleuth/docs/current/reference/htmlsingle/#features-baggage

File: text/logs/0150-logging-library-sdk.md
Comment (tigrannajaryan) [reply text/logs/0150-logging-library-sdk.md]: Fixed, thanks.

File: text/logs/0150-logging-library-sdk.md
Comment (tigrannajaryan) [reply text/logs/0150-logging-library-sdk.md]: Fixed, thanks.

**Review by jkwatson — APPROVED** [review]
Looks like a good starting point to me.

File: text/logs/0150-logging-library-sdk.md
Comment (lalitb) [inline text/logs/0150-logging-library-sdk.md:60]: As part of current cpp [implementation ](https://github.com/open-telemetry/opentelemetry-cpp/blob/1e5751e20367b715221bf3820b84590559170cdf/sdk/include/opentelemetry/sdk/logs/logger_provider.h#L60), we are using logger-name instead of instrumentation library name to get the correct `Logger`. I understand there have been some discussion on including the logger-name in log-record data model, not sure the current status though.

File: text/logs/0150-logging-library-sdk.md
Comment (tigrannajaryan) [reply text/logs/0150-logging-library-sdk.md:78]: Thanks. I added this example to the " Open questions" section.

File: text/logs/0150-logging-library-sdk.md
Comment (lalitb) [inline text/logs/0150-logging-library-sdk.md:89]: OK. thanks for providing the LoggerName PR link here. Please ignore my earlier comment.

File: text/logs/0150-logging-library-sdk.md
Comment (tigrannajaryan) [reply text/logs/0150-logging-library-sdk.md:60]: Yes, https://github.com/open-telemetry/opentelemetry-specification/pull/1236 suggests that the instrumentation library name and logger name are distinct and both needed.

File: text/logs/0150-logging-library-sdk.md
Comment (tigrannajaryan) [reply text/logs/0150-logging-library-sdk.md:89]: Please feel free to comment on that PR.

**Review by lalitb — APPROVED** [review]
LGTM in general. Good start for initial prototyping.

**Review by owais — APPROVED** [review]
LGTM

**Review by ThomsonTan — APPROVED** [review]

**Review by kumoroku — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Has the necessary number of approvals. No new comments for 3 days. All comments resolved. Merging.

If anyone sees this PR after merging and has thoughts please do speak: we want to know your opinion and can adjust the OTEP and our prototyping approach.

**Pull request merged**
